src/circuit_proof/examples.rs

```diff
@@ -0,0 +1,247 @@
+#![allow(non_snake_case)]
+
+use super::r1cs::{ConstraintSystem, LinearCombination, Variable};
+use curve25519_dalek::scalar::Scalar;
+use errors::R1CSError;
+use generators::{Generators, PedersenGenerators};
+use proof_transcript::ProofTranscript;
+use rand::rngs::OsRng;
+
+// TODO: make a trait that all circuit examples need to implement
+
+// 2-in 2-out merge circuit
+pub struct Merge {}
+
+impl Merge {
+    pub fn new() -> Self {
+        Merge {}
+    }
+
+    pub fn make_prover_cs(
+        &self,
+        type_0: Scalar,
+        type_1: Scalar,
+        val_in_0: Scalar,
+        val_in_1: Scalar,
+        val_out_0: Scalar,
+        val_out_1: Scalar,
+    ) -> (ConstraintSystem, Scalar) {
```

Comment:
If the merge block is a circuit gadget, shouldn't it take a `&mut ConstraintSystem` as an input rather than returning one that it created? Since we don't have a way to combine constraint systems (and I don't think there's a good way to do that), doesn't constructing internally and returning the `ConstraintSystem` prevent joining two merge gadgets?

Reply:
yeah that definitely makes more sense - fixed!